From 1baeb2308abaacfee9bbf84d0499df22b8a9181c Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:06:16 -0400
Subject: [tools] perf: Adjust symbols in VDSO

Message-id: <1410181752-25631-132-git-send-email-jolsa@redhat.com>
Patchwork-id: 91777
O-Subject: [PATCH RHEL7.1 BZ1134356 131/307] perf tools: Adjust symbols in VDSO
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 922d0e4d9f042f68baea28e5b3bd7bb708875f2e
Author: Vladimir Nikulichev <nvs@tbricks.com>
Date: Thu Apr 17 08:27:01 2014 -0700

description
===========
pert-report doesn't resolve function names in VDSO:

$ perf report --stdio -g flat,0.0,15,callee --sort pid
...
            8.76%
               0x7fff6b1fe861
               __gettimeofday
               ACE_OS::gettimeofday()
...

In this case symbol values should be adjusted the same way as for executables,
relocatable objects and prelinked libraries.

After fix:

$ perf report --stdio -g flat,0.0,15,callee --sort pid
...
            8.76%
               __vdso_gettimeofday
               __gettimeofday
               ACE_OS::gettimeofday()
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/symbol-elf.c b/tools/perf/util/symbol-elf.c
index 3b7dbf5..6864661 100644
--- a/tools/perf/util/symbol-elf.c
+++ b/tools/perf/util/symbol-elf.c
@@ -6,6 +6,7 @@
 #include <inttypes.h>
 
 #include "symbol.h"
+#include "vdso.h"
 #include <symbol/kallsyms.h>
 #include "debug.h"
 
@@ -618,6 +619,7 @@ int symsrc__init(struct symsrc *ss, struct dso *dso, const char *name,
   GElf_Shdr shdr;
   ss->adjust_symbols = (ehdr.e_type == ET_EXEC ||
     ehdr.e_type == ET_REL ||
+    is_vdso_map(dso->short_name) ||
     elf_section_by_name(elf, &ehdr, &shdr,
            ".gnu.prelink_undo",
            NULL) != NULL);
-- 
1.7.1